### PR TITLE
fix(framework): added defaultToZeroIfEmpty prop

### PR DESCRIPTION
**Pull request by richardilleskosmala**
With defaultToZeroIfEmpty we can make sure the ComboPicker always has an input value, for the cases in which we don't want input value to be optional

Closes: 86963qkcg

### CODE REVIEW

File: framework/lib/components/text-field/formatted-number-input.tsx [hunk 1]
```diff
@@ -75,28 +76,41 @@ export const FormattedNumberInput = ({
...
   max = Infinity,
   min = -Infinity,
   inputLeftElement,
   inputRightElement,
+  onBlur,
   ...rest
 }: FormattedNumberInputProps) => {
+  const [ v, setV ] = useState(valueProp)
+  const isControlled = typeof valueProp !== 'undefined'
+  const value = isControlled ? valueProp : v
```
Comment (SebCodesTheWeb) [inline framework/lib/components/text-field/formatted-number-input.tsx:90]: Okay so I'm very nitpicky here and I know the internal state was called "v" before but could you rename it to valueState? So the distinction between valueState and valueProp is clear

File: framework/lib/components/combo-picker/combo-picker.tsx [hunk 2]
```diff
@@ -30,15 +34,24 @@ export const ComboPicker = ({
...
+    (is(Number, input))
+      ? { input: Number(input), option }
+      : { option })
+
+  const getInputWithinRange = ((input: number) => {
+    if (max && input > max) return max
+    if (min && input < min) return min
+
+    return input
+  })
```
Comment (SebCodesTheWeb) [inline framework/lib/components/combo-picker/combo-picker.tsx:47]: Default max=Infinity and min=-Infinity and then return clamp(min, max, input), it's a util in northlight

File: framework/lib/components/combo-picker/combo-picker.tsx [hunk 3]
```diff
@@ -78,13 +91,20 @@ export const ComboPicker = ({
       <FormattedNumberInput
         width="100%"
         onChange={ handleInputChange }
-        defaultValue={ inputValue }
+        value={ inputValue }
```
Comment (SebCodesTheWeb) [inline framework/lib/components/combo-picker/combo-picker.tsx:94]: Hey what happens if I try to change the value prop from outside, would it react and update the internal state of combo-picker?

File: framework/lib/components/combo-picker/combo-picker.tsx [hunk 4]
```diff
@@ -78,13 +91,20 @@ export const ComboPicker = ({
...
         numberOfDecimals={ precision }
         preset={ formatPreset }
         disabled={ isDisabled }
         readOnly={ isReadOnly }
+        min={ min }
+        max={ max }
+        onBlur={ () => {
+          if (!is(Number, inputValue) && defaultToZeroIfEmpty) {
+            setInputValue(0)
+          }
```
Comment (SebCodesTheWeb) [inline framework/lib/components/combo-picker/combo-picker.tsx:106]: Check if inputValue isNil from ramda instead, it should be more direct?

File: framework/lib/components/combo-picker/combo-picker.tsx [hunk 5]
```diff
@@ -30,15 +34,24 @@ export const ComboPicker = ({
...
+    return input
+  })
 
   const handleInputChange = (newInputvalue: NumberFormatValues) => {
     const newValue = getNewValue(selectOption, newInputvalue.floatValue)
 
     setInputValue(newValue.input)
-    onChange?.(newValue)
+    onChange?.({ ...newValue,
+      ...!is(Number, newValue.input) && defaultToZeroIfEmpty && { input: getInputWithinRange(0) } })
```
Comment (SebCodesTheWeb) [inline framework/lib/components/combo-picker/combo-picker.tsx:54]: Maybe put this in a separate variable? `const clampedDefaultValue = ...`

**Review by SebCodesTheWeb — CHANGES_REQUESTED** [review]

File: framework/lib/components/combo-picker/combo-picker.tsx (hunk 4 above)
Comment (richardilleskosmala) [reply framework/lib/components/combo-picker/combo-picker.tsx:106]: If the inputValue would be NaN for some reason, it would not be properly handled by isNil, so for all the checks on this component and MoneyInputField I'm just checking if it's a number instead, and if it's not a number than it's not valid

File: framework/lib/components/text-field/formatted-number-input.tsx (hunk 1 above)
Comment (richardilleskosmala) [reply framework/lib/components/text-field/formatted-number-input.tsx]: Done!

File: framework/lib/components/combo-picker/combo-picker.tsx (hunk 2 above)
Comment (richardilleskosmala) [reply framework/lib/components/combo-picker/combo-picker.tsx]: Changed

File: framework/lib/components/combo-picker/combo-picker.tsx (hunk 5 above)
Comment (richardilleskosmala) [reply framework/lib/components/combo-picker/combo-picker.tsx]: Done!

File: framework/lib/components/combo-picker/combo-picker.tsx (hunk 3 above)
Comment (richardilleskosmala) [reply framework/lib/components/combo-picker/combo-picker.tsx]: Fixed

File: framework/lib/components/combo-picker/combo-picker.tsx [hunk 6]
```diff
@@ -14,30 +15,40 @@ export const ComboPicker = ({
...
   ...rest
 }: ComboPickerProps) => {
   const { isOpen, onToggle, onClose } = useDisclosure()
-  const [ inputValue, setInputValue ] = useState(value?.input)
-  const [ selectOption, setSelectOption ] = useState(value?.option ?? options[0])
+  const [ inputValueState, setInputValueState ] = useState(valueProp?.input)
+  const [ selectOptionState, setselectOptionState ] = useState(valueProp?.option ?? options[0])
   const [ enableSelectInput, setEnableSelectInput ] = useState(false)
 
+  const isInputValueControlled = typeof valueProp !== 'undefined'
```
Comment (SebCodesTheWeb) [inline framework/lib/components/combo-picker/combo-picker.tsx:34]: isInputValueControlled and isOptionControlled are the same here?

File: framework/lib/components/combo-picker/combo-picker.tsx [hunk 7]
```diff
@@ -73,18 +86,26 @@ export const ComboPicker = ({
     }
   }, [ enableSelectInput ])
 
+  useEffect(() => {
+    if (!is(Number, inputValue) && defaultToZeroIfEmpty) {
+      handleInputChange(clamp(min, max, 0))
+    }
+  }, [ inputValue ])
```
Comment (SebCodesTheWeb) [inline framework/lib/components/combo-picker/combo-picker.tsx:93]: Hmmm handleInputChange calls setInputValueState, which means that if the value is uncontrolled this could turn into a loop, maybe the useEffect dependency should only be valueProp.input?

File: framework/lib/components/text-field/formatted-number-input.tsx [hunk 8]
```diff
@@ -1,10 +1,11 @@
-import React, { ComponentType, useState } from 'react'
+import React, { ComponentType, useEffect, useState } from 'react'
 import { identity, isNil } from 'ramda'
 import {
   InputAttributes,
   NumberFormatValues,
```
Comment (SebCodesTheWeb) [inline framework/lib/components/text-field/formatted-number-input.tsx:5]: This changes for formatted-number-input could be put in a separate commit

File: framework/lib/components/text-field/formatted-number-input.tsx [hunk 9]
```diff
@@ -75,28 +76,41 @@ export const FormattedNumberInput = ({
...
     const factor = isPercentage ? 100 : 1
     if (vNum * factor > max) {
       const newValue = roundToPrecision(max / factor, numberOfDecimals)
-      setV(newValue)
+      setValueState(newValue)
+      onChange({
+        floatValue: newValue,
+        formattedValue: numericFormatter(newValue.toString(), props),
+        value: newValue.toString(),
+      })
```
Comment (SebCodesTheWeb) [inline framework/lib/components/text-field/formatted-number-input.tsx:104]: Maybe put in separate function, getNumberFormatValueObject or similiar? This is a preference detail tho so feel free to skip

**Review by SebCodesTheWeb — CHANGES_REQUESTED** [review]

**Review by SebCodesTheWeb — APPROVED** [review]
